### PR TITLE
aliases Property for Cfgu Declaration

### PR DESCRIPTION
**Pull request by Aman1919**
related to https://github.com/configu/configu/issues/239

### CODE REVIEW

File: ts/packages/ts/src/commands/EvalCommand.ts [hunk 1]
```diff
@@ -255,6 +255,15 @@ export class EvalCommand extends Command<EvalCommandReturn> {
             ERR(`one or more depends of key "${key}" is missing a value`, { location: [`EvalCommand`, 'run'] }),
           );
         }
+
+        if (evaluatedValue && cfgu.aliases && !cfgu.aliases.some((alias) => alias === key)) {
```
Comment (RonConfigu) [inline ts/packages/ts/src/commands/EvalCommand.ts:259]: There are few things i don't understand here:
- Why you feel like `validate` is the right place for aliases? EvalCommand can run without validation if needed. it's in its props.
- This condition forces me to have one of my aliases as the name of the config. This is not one of the requirements from the issue.
- even if i'll do as this condition expect, i'd still won't have the aliases in the result from EvalCommand. see attached
```javascript
// Schema
schema = {
  "MY_EMAIL_ID": {
    "type": "String",
    "default": "1",
    "aliases": [
      "emailId", "MY_EMAIL_ID"
    ]
  }
}
// EvalCommand result
result = { MY_EMAIL_ID: '1' }

// Expected result:
result = { MY_EMAIL_ID: '1', emailId: '1' }
```

File: ts/packages/ts/src/commands/UpsertCommand.ts [hunk 2]
```diff
@@ -31,6 +31,7 @@ export class UpsertCommand extends Command<void> {
       .map<Config>(([key, value]) => {
         const cfgu = schemaContents[key];
 
+        let AllAliases: string[] = [];
```
Comment (RonConfigu) [inline ts/packages/ts/src/commands/UpsertCommand.ts:34]: Why would create this here? for every config?
If at-all this should be declared before the loop

File: ts/packages/ts/src/commands/UpsertCommand.ts [hunk 3]
```diff
@@ -66,6 +67,25 @@ export class UpsertCommand extends Command<void> {
             }),
           );
         }
+        if (value && cfgu.aliases) {
+          if (AllAliases.some((a) => a === key)) {
```
Comment (RonConfigu) [inline ts/packages/ts/src/commands/UpsertCommand.ts:71]: This check will **always** be false.

File: ts/packages/ts/src/commands/UpsertCommand.ts [hunk 4]
```diff
@@ -66,6 +67,25 @@ export class UpsertCommand extends Command<void> {
...
+            throw new Error(
+              ERR(`invalid config  key "${key}"`, {
+                location: scopeLocation,
+                suggestion: `upserting more than one of any of the above is invalid.`,
+              }),
+            );
+          }
+          if (cfgu.aliases.some((a) => a === key)) {
```
Comment (RonConfigu) [inline ts/packages/ts/src/commands/UpsertCommand.ts:79]: This check's logic is opposite to the check your'e doing in EvalCommand line 259.
Can you explain the thought process?

File: ts/packages/ts/src/ConfigSchema.ts [hunk 5]
```diff
@@ -16,7 +16,18 @@ export abstract class ConfigSchema implements IConfigSchema {
...
+      'type',
+      'pattern',
+      'schema',
+      'default',
+      'required',
+      'depends',
+      'template',
+      'description',
+      'aliases',
+      'options',
```
[the remark this answers is not in the document]
Comment (Aman1919) [reply ts/packages/ts/src/ConfigSchema.ts:29]: ok next time will a issue

File: ts/packages/ts/src/ConfigSchema.ts [hunk 6]
```diff
@@ -325,13 +336,13 @@ export abstract class ConfigSchema implements IConfigSchema {
   static async parse(schema: ConfigSchema) {
     const scopeLocation = [`ConfigSchema`, 'parse', schema.path];
     const contents = await schema.read();
+    const AllAliases: string[] = [];
```
[the remark this answers is not in the document]
Comment (Aman1919) [reply ts/packages/ts/src/ConfigSchema.ts]: I changeI changed it to set

File: ts/packages/ts/src/ConfigSchema.ts [hunk 7]
```diff
@@ -395,6 +406,56 @@ export abstract class ConfigSchema implements IConfigSchema {
...
+            if (alias && !NAME(alias)) {
+              throw new Error(
+                ERR(`invalid aliases property`, {
+                  location: [...scopeLocation, `${key}[${idx}]`],
+                  suggestion: `alias mustn't contain reserved words "${key}"`,
+                }),
+              );
+            }
+
+            if (alias && !ConfigSchema.CFGU.TESTS.VAL_TYPE[type]?.({ ...cfgu, value: alias })) {
```
[the remark this answers is not in the document]
Comment (Aman1919) [reply ts/packages/ts/src/ConfigSchema.ts]: removed it .

File: ts/packages/ts/src/commands/EvalCommand.ts (hunk 1 above)
Comment (Aman1919) [reply ts/packages/ts/src/commands/EvalCommand.ts]: removed it

File: ts/packages/ts/src/ConfigSchema.ts
[the remark this answers is not in the document]
Comment (Aman1919) [reply ts/packages/ts/src/ConfigSchema.ts:1]: for this i was using   _(schemaContents).keys().value() lodash to get all the keys and on line 438  in configSchema it checks if any key is  equal to any of the aliases

File: ts/packages/ts/src/commands/UpsertCommand.ts (hunk 4 above)
Comment (Aman1919) [reply ts/packages/ts/src/commands/UpsertCommand.ts]: for eval i created a function which will iterate result and  if it keys has aliases then it add its alias with it value

**Comment by Aman1919** [comment]
> Hi @Aman1919,
> 
> Your contributions to our open source project are deeply appreciated, and they play a crucial role in strengthening our community. 💯 🥇
> 
> I've noticed some of the task's requirements are not implemented. EvalCommand result for example.
> I encourage you to review my comments and also take a moment to read @RichardAkman's [comment](https://github.com/configu/configu/issues/239#issuecomment-1770874257). I believe his insight will provide you with a better understanding of the task requirements.
> 
> If you have any questions or need guidance, please don't hesitate to reach out to us on Discord.
> 
> Best regards,
> Ron Roditi

@RonConfigu I did some changes can u review it

**Comment by RonConfigu** [comment]
> > Hi @Aman1919,
> > Your contributions to our open source project are deeply appreciated, and they play a crucial role in strengthening our community. 💯 🥇
> > I've noticed some of the task's requirements are not implemented. EvalCommand result for example.
> > I encourage you to review my comments and also take a moment to read @RichardAkman's [comment](https://github.com/configu/configu/issues/239#issuecomment-1770874257). I believe his insight will provide you with a better understanding of the task requirements.
> > If you have any questions or need guidance, please don't hesitate to reach out to us on Discord.
> > Best regards,
> > Ron Roditi
> 
> @RonConfigu I did some changes can u review it . i apologize that there were so many errors

Hi @Aman1919,
Thank you for your message, and I appreciate your efforts in making changes. Mistakes happen, and it’s all part of the learning process. :blush:
However, it seems that the branch still emits errors, and some of the task requirements are not yet met. Your potential is evident, and I’m confident you can resolve this.
If you have any questions or need guidance, please don’t hesitate to ask. We’re here to support you in your contributions.
Best regards,
Ron Roditi

**Review by RonConfigu — CHANGES_REQUESTED** [review]
Hi @Aman1919,
Thank you for your message, and I appreciate your efforts in making changes. Mistakes happen, and it’s all part of the learning process. :blush:
However, it seems that the branch still emits errors, and some of the task requirements are not yet met. Your potential is evident, and I’m confident you can resolve this.
If you have any questions or need guidance, please don’t hesitate to ask. We’re here to support you in your contributions.
Best regards,
Ron Roditi

**Pull request closed without merging**